### PR TITLE
Add project, "event-reader-sink"

### PR DESCRIPTION
**Pull request by mtammineni**

### CODE REVIEW

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/camel/AwsRoutes.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package gov.cdc.ocio.eventreadersink.camel
+
+import org.apache.camel.builder.RouteBuilder
+import gov.cdc.ocio.eventreadersink.model.AwsConfig
+import mu.KotlinLogging
+import org.apache.camel.LoggingLevel
+
+private val logger = KotlinLogging.logger {}
+
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/camel/AwsRoutes.kt:9]: Class header description missing

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/camel/AwsRoutes.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+class AwsRoutes(private val awsConfig: AwsConfig) : RouteBuilder() {
+    override fun configure() {
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/camel/AwsRoutes.kt:11]: Function header missing

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/cloud/AwsManager.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            sqsConfiguration.secretKey = awsSecretAccessKey
+            sqsConfiguration.region = awsRegion
+            sqsComponent.configuration = sqsConfiguration
+            return sqsComponent
+        } catch (e: IllegalArgumentException) {
+            logger.error("Invalid argument provided for SQS configuration: ${e.message}")
+            throw e
+        } catch (e: Exception) {
+            logger.error("Error configuring AWS SQS component: ${e.message}", e)
+            throw ConfigurationException("Failed to configure AWS SQS component: ${e.message}")
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/cloud/AwsManager.kt:45]: If you are throwing ConfigurationException then use that to catch and then have another catch block for catch "all" exception

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/cloud/AwsManager.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            s3Configuration.secretKey = awsSecretAccessKey
+            s3Configuration.region = awsRegion
+            s3Component.configuration = s3Configuration
+            return s3Component
+        } catch (e: IllegalArgumentException) {
+            logger.error("Invalid argument provided for S3 configuration: ${e.message}")
+            throw e
+        } catch (e: Exception) {
+            logger.error("Error configuring AWS S3 component: ${e.message}", e)
+            throw ConfigurationException("Failed to configure AWS S3 component: ${e.message}")
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/cloud/AwsManager.kt:78]: Same as above

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/cloud/CloudProviderType.kt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package gov.cdc.ocio.eventreadersink.cloud
+
+enum class CloudProviderType (val value: String) {
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/cloud/CloudProviderType.kt:3]: Class header missing

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/BadServiceException.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package gov.cdc.ocio.eventreadersink.exceptions
+
+/**
+ * Intended use of this exception is for bad requests, or, if a required parameter is missing for a request.
+ *
+ * @constructor
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/BadServiceException.kt:6]: @param missing

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/BadStateException.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package gov.cdc.ocio.eventreadersink.exceptions
+
+/**
+ * Intended use of this exception is for internal server issues where we expect to be in a certain state or have
+ * internal state information that is missing or invalid.
+ *
+ * @constructor
+ */
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/BadStateException.kt:8]: Same as above

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/ConfigurationException.kt [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package gov.cdc.ocio.eventreadersink.exceptions
+
+/**
+ * Exception thrown for general configuration loading errors.
+ */
+class ConfigurationException(message: String) : Exception(message)
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/ConfigurationException.kt:6]: Same as above

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/MissingPropertyException.kt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package gov.cdc.ocio.eventreadersink.exceptions
+
+/**
+ * Exception thrown when a required configuration property is missing.
+ */
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/MissingPropertyException.kt:5]: Same

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/model/CloudConfig.kt [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * This data class allows for the selection between AWS and Azure configurations,
+ * encapsulating the settings necessary for accessing the respective cloud services.
+ *
+ * @property provider The cloud provider to use ("aws" or "azure").
+ * @property awsConfig Optional configuration for AWS services, represented by an AwsConfig instance.
+ * @property azureConfig Optional configuration for Azure services, represented by an AzureConfig instance.
+ */
+data class CloudConfig(
+    var provider: String, // "aws" or "azure"
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/model/CloudConfig.kt:14]: Shold this be the enum defined above?

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AwsSink.kt [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            // Configure AWS Components
+            configureAwsComponents(awsConfig, camelContext)
+
+            // Add routes
+            camelContext.addRoutes(AwsRoutes(awsConfig))
+
+            // Start the Camel context
+            camelContext.start()
+            Runtime.getRuntime().addShutdownHook(Thread { camelContext.stop() })
+        } catch (e: ConfigurationException) {
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AwsSink.kt:52]: Did it correctly here :)

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AwsSink.kt [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            camelContext.addComponent("aws2-sqs", sqsComponent)
+
+            val s3Component = AwsManager().configureAwsS3Component(
+                awsConfig.accessKeyId,
+                awsConfig.secretAccessKey,
+                awsConfig.s3Region
+            )
+            camelContext.addComponent("aws2-s3", s3Component)
+
+        } catch (e: Exception) {
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AwsSink.kt:91]: Do as above

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                camelContext,
+                serviceBusNamespace,
+                sharedAccessKeyName,
+                sharedAccessKey
+            )
+            camelContext.addRoutes(AzureRoutes(topicName, subscriptionName, accountName, accountKey, containerName))
+            startCamelContext(camelContext)
+        } catch (e: Exception) {
+            logger.error("Failed to sink messages from Azure Service Bus to Blob Storage: ${e.message}", e)
+            throw BadServiceException("Error initializing AzureSink: ${e.message}")
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt:65]: Same issue - catch BaseServiceException and have another catch for Exception

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+     * @throws BadStateException
+     */
+    @Throws (BadStateException:: class)
+    private fun startCamelContext(camelContext: CamelContext) {
+        try {
+            camelContext.start()
+            Runtime.getRuntime().addShutdownHook(Thread {
+                try {
+                    camelContext.stop()
+                } catch (e: Exception) {
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt:81]: Same as above

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+                    logger.error("Failed to stop Camel context: ${e.message}", e)
+                    throw BadStateException("Error starting Camel context: ${e.message}")
+                }
+            })
+        } catch (e: Exception) {
+            logger.error("Failed to start Camel context: ${e.message}", e)
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt:87]: Same

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+            }
+
+            val amqpComponent = AMQPComponent().apply {
+                setCamelContext(camelContext)
+                this.connectionFactory = connectionFactory
+            }
+
+            camelContext.addComponent("amqp", amqpComponent)
+        } catch (e: Exception) {
+            logger.error("Failed to configure AMQP component: ${e.message}", e)
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt:124]: Same issue here

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/CamelProcessor.kt [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+                    }
+                }
+                else -> {
+                    logger.error ("Unsupported cloud provider: $provider")
+                    throw IllegalArgumentException("Unsupported cloud provider: $provider")
+                }
+            }
+        } catch (e: Exception) {
+            logger.error ("Error processing cloud configuration: ${e.message}" )
+            throw BadStateException("Error processing cloud configuration:  ${e.message}")
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/CamelProcessor.kt:78]: Same issue

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/EventProcessor.kt [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+            camelProcessor.sinkMessageToStorage(cloudConfig)
+        } catch (e: IllegalArgumentException) {
+            logger.error("Failed to process event due to invalid argument: ${e.message}", e)
+            throw BadStateException("Invalid argument provided: ${e.message}")
+        } catch (e: BadStateException) {
+            logger.error("Failed to process event: ${e.message}")
+            throw e // Re-throwing the original exception
+        } catch (e: Exception) {
+            logger.error("Failed to process event due to an unexpected error: ${e.message}", e)
+            throw BadStateException("Unexpected error occurred  ${e.message}")
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/EventProcessor.kt:41]: BadStateException is already caught why throwing again?

**Review by manu-govind — CHANGES_REQUESTED** [review]

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/camel/AwsRoutes.kt (hunk 1 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/camel/AwsRoutes.kt:9]: Updated.

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/cloud/CloudProviderType.kt (hunk 5 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/cloud/CloudProviderType.kt:12]: Updated class header

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/BadServiceException.kt (hunk 6 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/BadServiceException.kt]: Added @param and definition

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/ConfigurationException.kt (hunk 8 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/ConfigurationException.kt:8]: Added @param and definition

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/MissingPropertyException.kt (hunk 9 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/exceptions/MissingPropertyException.kt:7]: Added @param and definition

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/model/CloudConfig.kt (hunk 10 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/model/CloudConfig.kt:14]: Updated

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AwsSink.kt (hunk 12 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AwsSink.kt]: Done

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt (hunk 13 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt:65]: Updated.

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt (hunk 14 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt]: Updated

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt (hunk 15 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt]: Updated

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt (hunk 16 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/AzureSink.kt:133]: Updated

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/CamelProcessor.kt (hunk 17 above)
Comment (mtammineni) [reply event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/CamelProcessor.kt:78]: Done

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/Application.kt [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+            subscriptionName = environment.config.property("cloud.azure.service_bus.subscription_name").getString(),
+            containerName = environment.config.property("cloud.azure.blob_storage.container_name").getString(),
+            storageAccountKey = environment.config.property("cloud.azure.blob_storage.storage_account_key").getString(),
+            storageAccountName = environment.config.property("cloud.azure.blob_storage.storage_account_name").getString()
+        )
+
+        CloudConfig(provider = CloudProviderType.AZURE, azureConfig = azureConfig)
+    } catch (e: MissingPropertyException) {
+        logger.error("Error: Missing required Azure configuration property: ${e.message}")
+        throw ConfigurationException("Missing required Azure configuration property.")
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/Application.kt:116]: Is it throw e or ConfigurationException?

File: event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/EventProcessor.kt [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+     * the processing, logging an error message if an exception is
+     * thrown.
+     */
+    @Throws(BadStateException::class, IllegalArgumentException::class)
+    fun processEvent() {
+        try {
+            camelProcessor.sinkMessageToStorage(cloudConfig)
+        } catch (e: IllegalArgumentException) {
+            logger.error("Failed to process event due to invalid argument: ${e.message}", e)
+            throw BadStateException("Invalid argument provided: ${e.message}")
```
Comment (manu-govind) [inline event-reader-sink/src/main/kotlin/gov/cdc/ocio/eventreadersink/sink/EventProcessor.kt]: throw e or BadStateException?

**Review by manu-govind — APPROVED** [review]
Looks good

**Pull request merged**